changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,552 @@
+a LRU Cache of Hierarchy Keyrings.
+Which, while workable, is sub-optimal,
+and clearly makes the Hierarchy Keyring,
+in these conditions,
+"Hard to Use".
+
+The objective, with these changes,
+is to make the Hierarchy Keyring
+"Easy to Use" in a multiple KMS Relationship
+environment.
+
+## Summary
+
+Relevant [sim](https://sim.amazon.com/issues/CrypTool-5311).
+
+To improve the MPL Consumer data key caching experience,
+when using the Hierarchy Keyring,
+we need to allow caching across Key Stores/KMS Clients/KMS Keys.
+
+To facilitate Caching across Key Stores/KMS Clients/KMS Keys,
+we MUST break the Cryptographic Materials Cache (CMC)
+out of the Hierarchy Keyring.
+
+By allowing MPL Consumers to provide an already initialized CMC
+to the Hierarchy Keyring at construction,
+the CMC MAY cache Branch Keys protected by different
+KMS Relationships.
+
+This simplifies Multiple KMS Relationship MPL Consumers,
+as they do not need to stand up a LRU Cache of Hierarchy Keyrings.
+
+Instead, they may maintain one CMC.
+They still create a Hierarchy Keyring instance per KMS Relationship,
+and they MUST use the correct Keyring to retrieve material
+from the Cache.
+
+But they only need to maintain the common cache.
+
+In the future, the CachingCMM will be introduced to
+Crypto Tool's Dafny products.
+The CachingCMM and the Hierarchy Keyring both consume CMCs.
+Thus, it will be possible to provide a CMC to both a
+Hierarchy Keyring & a CachingCMM.
+This cache identifier design MUST consider
+how the cache entries of multiple Hierarchy Keyrings
+and CachingCMMs will be appropriately separated.
+
+## Requirements
+
+- We must allow caching across Key Stores/KMS Clients/KMS Keys
+  for multiple Hierarchy Keyrings and CachingCMMs.
+  For this, we need to strategically update the cache identifiers
+  for all materials ([background](#background)) in the CMC.
+  The [Issues and Alternatives section](#issues-and-alternatives)
+  discusses this.
+
+## Out of Scope
+
+- Supporting a shared Cache in the DB-ESDK to fetch
+  BeaconKeyMaterials across multiple Searchable
+  Encryption Configurations, or across multiple KMS Relationships.
+
+## Issues and Alternatives:
+
+Before looking at the issues, please take a quick look at
+[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
+[CachingCMM Cache Identifier formula](#cachingcmm-cache-identifier-formula), and
+[DB-ESDK Searchable Encryption Cache Identifier formula](#db-esdk-searchable-encryption-cache-identifier-formula) for more context.
+
+Note that there are two hashings taking place.
+For example in the Hierarchy Keyring Cache Identifier formula,
+one hash is for the branch-key-id and the other is hashing
+the entire appended string
+(a part of which is the hash of the branch-key-id)
+to get the final cache identifier.
+We'll call these internal and external hashes.
+
+#### Issue 1: How to update the cache identifier for the Hierarchy Keyring to allow a shared cache?
```

Comment:
nit/formatting: you went from two `#` to four `#`, instead of 3 `#`.
I like neat hierarchies, 
but this 100% an author's choice.

Reply:
I had another heading there that I removed in the subsequent draft. Thanks for pointing this out.